src/v/cluster/topic_table.cc

```diff
@@ -649,22 +649,16 @@ void topic_table::notify_waiters() {
     /// function_ptrs stored in \ref notifications, without consuming all
     /// pending_deltas for when a subsequent waiter does arrive
 
-    std::vector<delta> changes;
-
     // \ref notify_waiters is called after every apply. Hence for the most
     // part there should only be a few items towards the end of \ref
     // _pending_deltas that need to be sent to callbacks in \ref
     // notifications. \ref _last_consumed_by_notifier_offset allows us to
     // skip ahead to the items added since the last \ref notify_waiters
     // call.
-    auto starting_iter = _pending_deltas.begin()
+    auto starting_iter = _pending_deltas.cbegin()
                          + _last_consumed_by_notifier_offset;
 
-    std::copy_if(
-      starting_iter,
-      _pending_deltas.end(),
-      std::back_inserter(changes),
-      [this](const delta& d) { return d.offset > _last_consumed_by_notifier; });
+    std::span changes{starting_iter, _pending_deltas.cend()};
```

Comment:
Maybe we should iterate through the `_pending_deltas` and assert that `d.offset > _last_consumed_by_notifier` instead of assuming it. Alternatively the commit message should explain why it's always true now.

Reply:
I think the old code assumed offsets were increasing and if they weren't, we would skip notification of the listeners in the `_notifications` array for some changes (but `_waiters` would be notified). 

The new logic doesn't assume anything about the offsets now, it just sends whatever suffix of the `_pending_deltas` array hasn't already been sent to the notifier in a previous call and keeps track of the offset within the vector of that suffix.

So I'm not sure there's anything to assert here.

Reply:
`_pending_deltas` should be monotonically increasing if the offsets applied to the `topic_table` by the `controller_stm` are in order. I believe this to be the case since if out of order offsets are applied to the topic_table then it could end up in an incorrect state. I.e, if a delete topic and create topic command was swapped.

Reply:
I think with current solution we can remove `_last_consumed_by_notifier`, it is no longer needed as we use iterators to provide the same semantics, previous code didn't leverage the fact that deltas were sorted by offset while current does.